### PR TITLE
add iam role

### PR DESCRIPTION
**Pull request by e-d-o-c**
### What?

I have added IAM roles for ECS and also enabled Fargate on the ECS Module

### Why?

I am doing this because it will allow the cluster to use Fargate and Fargate will needs permission to launch container instances and register them into the cluster

### CODE REVIEW

File: aws/ecs-service/files/container-instance-assume-policy.json [hunk 1]
```diff
@@ -0,0 +1,13 @@
+{
+    "Version": "2008-10-17",
```
Comment (amberstarlight) [inline aws/ecs-service/files/container-instance-assume-policy.json:2]:
```suggestion
    "Version": "2012-10-17",
```

Can we use the [newest version](https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_elements_version.html) here?

File: aws/ecs-service/files/container-instance-assume-policy.json [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+    "Statement": [
+      {
+        "Sid": "",
```
Comment (amberstarlight) [inline aws/ecs-service/files/container-instance-assume-policy.json:5]:
```suggestion
```

We don't need a `sid` here (it's empty)

File: aws/ecs-service/iam.tf [hunk 3]
```diff
@@ -0,0 +1,47 @@
+/* 
+ *  IAM role for fargate instances to use when they are launched
+ */
+
+
+data "aws_iam_policy_document" "ecs_tasks_execution_role_policy" {
```
Comment (amberstarlight) [inline aws/ecs-service/iam.tf:6]:
```suggestion
data "aws_iam_policy_document" "ecs_assume_role_policy" {
```

I would probably be more explicit here, and call it the assume role policy - as it's not got any other policy statements.

File: aws/ecs-service/iam.tf [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    principals {
+      type        = "Service"
+      identifiers = ["ecs-tasks.amazonaws.com"]
+    }
+  }
+}
+
+
+resource "aws_iam_role" "ecs_tasks_execution_role" {
+  name               = "TradeTariff-${var.environment}-ecs-task-execution-role"
```
Comment (amberstarlight) [inline aws/ecs-service/iam.tf:19]: We'll run into naming clashes here as multiple tasks use the module, should we have this name generated using more input variables?

File: aws/ecs-service/iam.tf [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+  assume_role_policy = data.aws_iam_policy_document.ecs_tasks_execution_role_policy.json
+}
+
+
+resource "aws_iam_role_policy_attachment" "ecs_tasks_execution_role" {
+  role       = aws_iam_role.ecs_tasks_execution_role.name
+  policy_arn = aws_iam_role.ecs_tasks_execution_role.arn
```
Comment (amberstarlight) [inline aws/ecs-service/iam.tf:26]: I'm guessing this is supposed to be an ARN of a policy?

File: aws/ecs-service/iam.tf [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+ */
+resource "aws_iam_role_policy" "container_instances_policy" {
+  name   = var.container_instance_policy
+  role   = aws_iam_role.container_instances.id
+  policy = file("${path.module}/files/container-instance-assume-policy.json")
+}
+
+resource "aws_iam_role" "container_instances" {
+  name               = "TradeTariff-${var.environment}-container-instances-role"
+  assume_role_policy = aws_iam_role_policy.container_instances_policy.json
```
Comment (amberstarlight) [inline aws/ecs-service/iam.tf:41]:
```suggestion
  assume_role_policy = file("${path.module}/files/container-instance-assume-policy.json")
```

This is where you'd put your assume role policy.

File: aws/ssm/example/README.md [hunk 7]
```diff
@@ -29,7 +29,7 @@ Note that this example may create resources which cost money. Run `terraform des
 
 | Name | Version |
 |------|---------|
-| <a name="provider_random"></a> [random](#provider\_random) | 3.4.3 |
+| <a name="provider_random"></a> [random](#provider\_random) | >= 3.0.0 |
```
Comment (willfish) [inline aws/ssm/example/README.md:32]: What is provider_random? Random number generation?

File: aws/ecs-service/iam.tf [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+
+
+/* 
+ *  Container instance Iam role
+ */
+resource "aws_iam_role_policy" "container_instances_policy" {
+  name   = var.container_instance_policy
+  role   = aws_iam_role.container_instances.id
+  policy = file("${path.module}/files/container-instance-assume-policy.json")
+}
```
Comment (amberstarlight) [inline aws/ecs-service/iam.tf:37]: `policy` here is where you'd want [`AmazonECS_FullAccess`](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/security-iam-awsmanpol-amazonecs-full-access-migration.html)

**Review by amberstarlight — CHANGES_REQUESTED** [review]

File: aws/ecs-service/files/container-instance-assume-policy.json (hunk 1 above)
Comment (amberstarlight) [reply aws/ecs-service/files/container-instance-assume-policy.json:2]: It'll work, as we're not using any of the new policy syntax, but AWS recommends against using it; 

> `2008-10-17`. This was an earlier version of the policy language. You might see this version on older existing policies. Do not use this version for any new policies or when you update any existing policies.

This is something they haven't refined on the autogenerated policies from the console.

**Pull request closed without merging**